src/Ace/Halogen/Component.purs

```diff
@@ -1,75 +1,185 @@
+foreign import initialized :: Ref Boolean
+foreign import focused :: Ref String
+foreign import dataset
+  :: forall e. HTMLElement -> Eff (dom :: DOM |e) (StrMap String)
+
+completeFnFocused :: forall e. Eff (AceEffects e) (CompleteFn e)
+completeFnFocused = do
+  focusedKey <- readRef focused
+  mFns <- readRef completeFns
+  maybe (pure emptyCompleteFn) pure $ Sm.lookup focusedKey mFns
+  where
+  emptyCompleteFn :: CompleteFn e
+  emptyCompleteFn _ _ _ _ = pure []
+
+setAutocompleteResume
+  :: forall e. Maybe Autocomplete -> Editor -> Eff (AceEffects e) Unit
+setAutocompleteResume Nothing editor = do
+  Editor.setEnableBasicAutocompletion false editor
+  Editor.setEnableLiveAutocompletion false editor
+setAutocompleteResume (Just Basic) editor = do
+  Editor.setEnableLiveAutocompletion false editor
+  Editor.setEnableBasicAutocompletion true editor
+setAutocompleteResume (Just Live) editor = do
+  Editor.setEnableLiveAutocompletion true editor
+  Editor.setEnableBasicAutocompletion true editor
+
+enableAutocomplete :: forall e. Eff (AceEffects e) Unit
+enableAutocomplete = do
+  languageToolsInitialized <- readRef initialized
+  when (not languageToolsInitialized) do
+    completer <- Completer.mkCompleter globalCompleteFn
+    tools <- LanguageTools.languageTools
+    LanguageTools.addCompleter completer tools
+    writeRef initialized true
+  where
+  globalCompleteFn editor session position prefix cb = do
+    fn <- completeFnFocused
+    runAff (const $ cb Nothing) (cb <<< Just)
+      $ fn editor session position prefix
+
+genKey :: forall e. Eff (now :: Now, random :: RANDOM|e) String
+genKey = do
+  rn1 <- random
+  rn2 <- random
+  (Milliseconds time) <- nowEpochMilliseconds
+  pure $ show rn1 <> show time <> show rn2
+
+----------------------------------------------------------------------
+-- Types, State and other common halogen things
+----------------------------------------------------------------------
+data Autocomplete = Live | Basic
+
+type AceEffects e =
+  ( random :: RANDOM
+  , now :: Now
+  , ref :: REF
+  , ace :: ACE
+  , avar :: AVAR
+  , dom :: DOM
+  | e)
+
+data AceQuery a
+  = Init HTMLElement a
+  | GetText (String -> a)
+  | SetText String a
+  | SetAutocomplete (Maybe Autocomplete) a
+  | SetCompleteFn (forall e. CompleteFn e) a
+  | GetEditor (Maybe Editor -> a)
+
+type CompleteFn e =
+  Editor -> EditSession -> Position -> String
+  -> Aff (AceEffects e) (Array Completion)
+
+type AceState = { key :: Maybe String
+                , editor :: Maybe Editor
+                }
+
+initialAceState :: AceState
+initialAceState = { key: Nothing, editor: Nothing }
 
 -- | The Ace component.
 aceComponent
   :: forall g eff
-   . (MonadEff (ace :: ACE | eff) g)
+   . (MonadEff (AceEffects eff) g)
   => (Editor -> g Unit)
-  -> Maybe String
+  -> Maybe Autocomplete
   -> Component AceState AceQuery g
-aceComponent setup initialText = component render eval
+aceComponent setup resume = component render eval
```

Comment:
In fact we don't need `SetText` because there is `setup` already. This will change `Notebook.Cell.Ace` a bit but I suppose that `setup` was added to solve same problems.